packages/client/src/perma-cache.js

```diff
@@ -0,0 +1,396 @@
+    this.rateLimiter = rateLimiter || createRateLimiter()
+  }
+
+  /**
+   * @hidden
+   * @param {string} token
+   * @returns {Record<string, string>}
+   */
+  static headers(token) {
+    /* c8 ignore next 1 */
+    if (!token) throw new Error('missing token')
+    return {
+      Authorization: `Bearer ${token}`,
+      'X-Client': 'nftstorage.link/js',
+    }
+  }
+
+  /**
+   * @param {Service} service
+   * @param {string[]} urls
+   * @param {PutOptions} [options]
+   * @returns {Promise<CacheResult[]>}
+   */
+  static async put(
+    { endpoint, token, rateLimiter = globalRateLimiter },
+    urls,
+    { onPut, maxRetries = MAX_RETRIES } = {}
+  ) {
+    areNftStorageLinkUrls(urls)
+
+    const headers = PermaCache.headers(token)
+
+    /** @type {import('p-settle').PromiseResult<CacheResult>[]} */
+    const cacheResults = await pSettle(
+      urls.map(async (url) => {
+        const apiUrl = new URL(
+          `perma-cache/${encodeURIComponent(url)}`,
+          endpoint
+        )
+
+        let cacheResult
+        try {
+          cacheResult = await pRetry(
+            async () => {
+              await rateLimiter()
+              const request = await fetch(apiUrl.toString(), {
+                method: 'POST',
+                headers,
+              })
+
+              /** @type {CacheResult} */
+              const res = await request.json()
+              if (!request.ok) {
+                // @ts-ignore Only exists on Error
+                const e = new Error(res.message)
+                // do not retry if abortable errors - will not succeed
+                if (ABORTABLE_ERRORS_STATUS.includes(request.status)) {
+                  throw new AbortError(e)
+                }
+                /* c8 ignore next 2 */
+                throw e
+              }
+
+              onPut && onPut(url)
+
+              return {
+                ...res,
+                url, // decoded
+              }
+            },
+            /* c8 ignore next 3 */
+            {
+              retries: maxRetries == null ? MAX_RETRIES : maxRetries,
+            }
+          )
+        } catch (error) {
+          return {
+            // @ts-ignore non error type
+            error: error.message,
+            url,
+          }
+        }
+        return cacheResult
+      })
+    )
+
+    // @ts-ignore p-settle will always handle error internally
+    return cacheResults.map((cacheResult) => cacheResult.value)
+  }
+
+  /**
+   * @param {Service} service
+   * @param {ListOptions} [options]
+   * @returns {AsyncIterable<CacheEntry>}
+   */
+  static async *list(
+    { endpoint, token, rateLimiter = globalRateLimiter },
+    { sort = 'date', order = 'asc' } = {}
+  ) {
+    const headers = PermaCache.headers(token)
+    let search = new URLSearchParams({ sort, order })
+    let nextPageUrl = new URL(`perma-cache?${search}`, endpoint)
+    do {
+      await rateLimiter()
+
+      const request = await fetch(nextPageUrl.toString(), {
```

Comment:
Retry this also?

Reply:
Well, this should not really fail. The critical one is put because of downloading content... We can add it here too, but I think we should go without it and rethink if it is a problem. Thoughts?